collector/pg_replication_slots.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func init() {
+	registerCollector("replication_slot", defaultEnabled, NewPGReplicationSlotCollector)
+}
+
+type PGReplicationSlotCollector struct {
+	log log.Logger
+}
+
+func NewPGReplicationSlotCollector(logger log.Logger) (Collector, error) {
+	return &PGReplicationSlotCollector{log: logger}, nil
+}
+
+var pgReplicationSlot = map[string]*prometheus.Desc{
+	"current_wal_lsn": prometheus.NewDesc(
+		"pg_replication_slot_current_wal_lsn",
+		"current wal lsn value",
+		[]string{"slot_name"}, nil,
+	),
+	"confirmed_flush_lsn": prometheus.NewDesc(
+		"pg_replication_slot_confirmed_flush_lsn",
+		"last lsn confirmed flushed to the replication slot",
+		[]string{"slot_name"}, nil,
+	),
+}
+
+func (PGReplicationSlotCollector) Update(ctx context.Context, db *sql.DB, ch chan<- prometheus.Metric) error {
+	rows, err := db.QueryContext(ctx,
+		`SELECT
+			slot_name,
+			pg_current_wal_lsn() AS current_wal_lsn,
+			confirmed_flush_lsn
+			FROM
+		pg_replication_slots;`)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var slot_name string
+		var wal_lsn int64
+		var flush_lsn int64
+		if err := rows.Scan(&slot_name, &wal_lsn, &flush_lsn); err != nil {
+			return err
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			pgReplicationSlot["current_wal_lsn"],
+			prometheus.GaugeValue, float64(wal_lsn), slot_name,
```

Comment:
Should these be a gauge? Won't this value only increase?

Reply:
counter sounds good to me. something i think i also need to do is that these values are in pg_lsn. im thinking in the sql statement do the `pg_lsn - - '0/0'` to get an actual int value. now for the null i can do `coalesce(confirmed_flush_lsn, '0/0') - '0/0'`

Reply:
Depends on if this is a truly a counter in the underlying system.